# Investigating WeRateDogs‘s Tweeter account
by Emadeldin Fawaz

## Installing needed packages and tweepy

In [1]:
import tweepy
import pandas as pd
import numpy as np
import requests
import json
import time
import matplotlib.pyplot as plt

In [1]:
# Tweepy keys and token

consumer_key = ''
consumer_secret = ''
access_token = ''
access_secret = ''

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth)

# 1. Data Gathering

In [4]:
# Supplied file
tw_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [5]:
# Requesting tweet image predictions

with open('image_predictions.tsv' , 'wb') as file:
    image_predictions = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv', auth=('user', 'pass'))
    file.write(image_predivtions.content)   

In [6]:
# Reading image predictions
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')

## Testing the data gathering process so far

In [6]:
tw_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [5]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [4]:
image_predictions.sample(5)

tweet_id                                          jpg_url  \
526   676613908052996102  https://pbs.twimg.com/media/CWPQwmJWUAAu_At.jpg   
945   704480331685040129  https://pbs.twimg.com/media/CcbRIAgXIAQaKHQ.jpg   
1356  760539183865880579  https://pbs.twimg.com/media/Co36VZfWcAEN3R3.jpg   
1420  771770456517009408  https://pbs.twimg.com/media/CrXhIqBW8AA6Bse.jpg   
2049  887101392804085760  https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   

      img_num           p1   p1_conf  p1_dog          p2   p2_conf  p2_dog  \
526         1  book_jacket  0.493790   False    Doberman  0.096423    True   
945         1      Samoyed  0.979206    True  Pomeranian  0.007185    True   
1356        1      Samoyed  0.988013    True    malamute  0.004518    True   
1420        1     papillon  0.533180    True      collie  0.192031    True   
2049        1      Samoyed  0.733942    True  Eskimo_dog  0.035029    True   

                               p3   p3_conf  p3_dog  
526            miniature_pinscher  0.070647    True  
945                    Arctic_fox  0.006438   False  
1356  West_Highland_white_terrier  0.001189    True  
1420                Border_collie  0.121626    True  
2049    Staffordshire_bullterrier  0.029705    True

In [12]:
# collecting data using API  

collected =[]
not_collected = [] # to avoid error/exception messages

with open('tweet_json.txt', 'w') as file:
    for tweet_id in list(tw_archive['tweet_id']):
        try:
            tweet_status = api.get_status(tweet_id,tweet_mode='extended')
            json.dump(tweet_status._json, file)
            file.write('\n')
            collected.append(tweet_id)
        except Exception as e:
            not_collected.append(tweet_id)
    
    

In [13]:
#Reading JSON content as pandas dataframe
tweet_status = pd.read_json('tweet_json.txt', lines = True,encoding='utf-8')

In [3]:
tweet_status.sample(5)

contributors  coordinates          created_at display_text_range  \
138           NaN          NaN 2017-05-15 19:14:50            [0, 93]   
513           NaN          NaN 2016-12-15 17:23:04            [0, 41]   
25            NaN          NaN 2017-07-17 16:17:36           [0, 101]   
688           NaN          NaN 2016-10-12 19:24:27            [0, 87]   
638           NaN          NaN 2016-10-31 02:17:31           [0, 104]   

                                              entities  \
138  {'hashtags': [], 'symbols': [], 'user_mentions...   
513  {'hashtags': [], 'symbols': [], 'user_mentions...   
25   {'hashtags': [], 'symbols': [], 'user_mentions...   
688  {'hashtags': [], 'symbols': [], 'user_mentions...   
638  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                     extended_entities  favorite_count  \
138  {'media': [{'id': 864197387966791680, 'id_str'...           30303   
513  {'media': [{'id': 809448667128074240, 'id_str'...            7438   
25   {'media': [{'id': 886983218871902208, 'id_str'...           34659   
688                                                NaN            8576   
638  {'media': [{'id': 792913326326026240, 'id_str'...           15442   

     favorited                                          full_text  geo  \
138      False  This is Paisley. She ate a flower just to prov...  NaN   
513      False  I call this one "A Blep by the Sea" 12/10 http...  NaN   
25       False  This is Maya. She's very shy. Rarely leaves he...  NaN   
688      False  This is Arnie. He's afraid of his own bark. 12...  NaN   
638      False  Here is a perfect example of someone who has t...  NaN   

                           ...                          quoted_status  \
138                        ...                                    NaN   
513                        ...                                    NaN   
25                         ...                                    NaN   
688                        ...                                    NaN   
638                        ...                                    NaN   

     quoted_status_id quoted_status_id_str  quoted_status_permalink  \
138               NaN                  NaN                      NaN   
513               NaN                  NaN                      NaN   
25                NaN                  NaN                      NaN   
688               NaN                  NaN                      NaN   
638               NaN                  NaN                      NaN   

     retweet_count  retweeted  retweeted_status  \
138           9072      False               NaN   
513           1618      False               NaN   
25            7652      False               NaN   
688           3152      False               NaN   
638           4439      False               NaN   

                                                source truncated  \
138  <a href="http://twitter.com/download/iphone" r...     False   
513  <a href="http://twitter.com/download/iphone" r...     False   
25   <a href="http://twitter.com/download/iphone" r...     False   
688  <a href="http://vine.co" rel="nofollow">Vine -...     False   
638  <a href="http://twitter.com/download/iphone" r...     False   

                                                  user  
138  {'id': 4196983835, 'id_str': '4196983835', 'na...  
513  {'id': 4196983835, 'id_str': '4196983835', 'na...  
25   {'id': 4196983835, 'id_str': '4196983835', 'na...  
688  {'id': 4196983835, 'id_str': '4196983835', 'na...  
638  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 32 columns]

## Data Assessment

In [2]:
# Reading gathered files
tw_archive = pd.read_csv('twitter-archive-enhanced.csv')
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')
tweet_status = pd.read_json('tweet_json.txt', lines = True,encoding='utf-8')

### Visual assessment

In [7]:
tw_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [10]:
image_predictions

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [13]:
tweet_status

contributors  coordinates          created_at display_text_range  \
0             NaN          NaN 2017-08-01 16:23:56            [0, 85]   
1             NaN          NaN 2017-08-01 00:17:27           [0, 138]   
2             NaN          NaN 2017-07-31 00:18:03           [0, 121]   
3             NaN          NaN 2017-07-30 15:58:51            [0, 79]   
4             NaN          NaN 2017-07-29 16:00:24           [0, 138]   
5             NaN          NaN 2017-07-29 00:08:17           [0, 138]   
6             NaN          NaN 2017-07-28 16:27:12           [0, 140]   
7             NaN          NaN 2017-07-28 00:22:40           [0, 118]   
8             NaN          NaN 2017-07-27 16:25:51           [0, 122]   
9             NaN          NaN 2017-07-26 15:59:51           [0, 133]   
10            NaN          NaN 2017-07-26 00:31:25           [0, 130]   
11            NaN          NaN 2017-07-25 16:11:53           [0, 107]   
12            NaN          NaN 2017-07-25 01:55:32           [0, 106]   
13            NaN          NaN 2017-07-25 00:10:02            [0, 91]   
14            NaN          NaN 2017-07-24 17:02:04           [0, 118]   
15            NaN          NaN 2017-07-24 00:19:32           [0, 138]   
16            NaN          NaN 2017-07-23 00:22:39            [0, 86]   
17            NaN          NaN 2017-07-22 16:56:37           [0, 128]   
18            NaN          NaN 2017-07-22 00:23:06            [0, 87]   
19            NaN          NaN 2017-07-20 16:49:33           [0, 127]   
20            NaN          NaN 2017-07-19 16:06:48           [0, 127]   
21            NaN          NaN 2017-07-19 03:39:09           [0, 108]   
22            NaN          NaN 2017-07-19 00:47:34            [0, 99]   
23            NaN          NaN 2017-07-18 16:08:03            [0, 88]   
24            NaN          NaN 2017-07-18 00:07:08           [0, 129]   
25            NaN          NaN 2017-07-17 16:17:36           [0, 101]   
26            NaN          NaN 2017-07-16 23:58:41           [0, 121]   
27            NaN          NaN 2017-07-16 20:14:00            [0, 71]   
28            NaN          NaN 2017-07-15 23:25:31           [0, 131]   
29            NaN          NaN 2017-07-15 16:51:35          [27, 105]   
..            ...          ...                 ...                ...   
742           NaN          NaN 2016-09-22 01:54:34           [0, 102]   
743           NaN          NaN 2016-09-22 01:16:45            [0, 95]   
744           NaN          NaN 2016-09-22 00:13:04            [0, 85]   
745           NaN          NaN 2016-09-21 17:42:10           [0, 121]   
746           NaN          NaN 2016-09-21 16:00:17            [0, 73]   
747           NaN          NaN 2016-09-21 01:39:11            [0, 73]   
748           NaN          NaN 2016-09-21 00:53:04           [0, 140]   
749           NaN          NaN 2016-09-21 00:00:35            [0, 99]   
750           NaN          NaN 2016-09-20 17:36:50           [0, 109]   
751           NaN          NaN 2016-09-20 01:12:28            [0, 68]   
752           NaN          NaN 2016-09-20 00:24:34           [0, 137]   
753           NaN          NaN 2016-09-19 19:31:59           [0, 140]   
754           NaN          NaN 2016-09-19 15:00:20           [0, 116]   
755           NaN          NaN 2016-09-19 01:42:24            [0, 91]   
756           NaN          NaN 2016-09-18 22:54:18           [0, 102]   
757           NaN          NaN 2016-09-18 21:33:11           [0, 112]   
758           NaN          NaN 2016-09-17 16:57:35            [0, 77]   
759           NaN          NaN 2016-09-16 16:24:19            [0, 84]   
760           NaN          NaN 2016-09-16 16:00:31           [0, 116]   
761           NaN          NaN 2016-09-15 17:48:25           [0, 103]   
762           NaN          NaN 2016-09-15 02:42:54           [0, 140]   
763           NaN          NaN 2016-09-15 00:36:55           [0, 104]   
764           NaN          NaN 2016-09-14 23:30:38            [

### Programmatic assessment

In [8]:
tw_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [9]:
tw_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [37]:
tw_archive[tw_archive['rating_denominator'] == 0 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313  835246439529840640           8.352460e+17           26259576.0   

                     timestamp  \
313  2017-02-24 21:54:03 +0000   

                                                source  \
313  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
313  @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp expanded_urls  \
313                       NaN                        NaN           NaN   

     rating_numerator  rating_denominator  name doggo floofer pupper puppo  
313               960                   0  None  None    None   None  None

In [38]:
tw_archive['text'][tw_archive['rating_denominator'] == 0 ]

313    @jonnysun @Lin_Manuel ok jomny I know you're e...
Name: text, dtype: object

In [ ]:
# delete this row.

In [39]:
tw_archive[tw_archive['rating_denominator'] < 10 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640           8.352460e+17           26259576.0   
516   810984652412424192                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

                      timestamp  \
313   2017-02-24 21:54:03 +0000   
516   2016-12-19 23:06:23 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                 source  \
313   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
2335  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
313   @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                  NaN   
2335  This is an Albanian 3 1/2 legged  Episcopalian...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
313                        NaN                        NaN   
516                        NaN                        NaN   
2335                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
313                                                 NaN               960   
516   https://www.gofundme.com/sams-smile,https://tw...                24   
2335  https://twitter.com/dog_rates/status/666287406...                 1   

      rating_denominator  name doggo floofer pupper puppo  
313                    0  None  None    None   None  None  
516                    7   Sam  None    None   None  None  
2335                   2    an  None    None   None  None

In [40]:
tw_archive[tw_archive['rating_denominator'] > 20 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
433   820690176645140481                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   

                      timestamp  \
433   2017-01-15 17:52:40 +0000   
902   2016-07-28 01:00:57 +0000   
1120  2016-05-13 16:15:54 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   

                                                 source  \
433   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
1120  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
1843  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
433   The floofs have been released I repeat the flo...                  NaN   
902   Why does this never happen at my front door......                  NaN   
1120  Say hello to this unbelievably well behaved sq...                  NaN   
1202  This is Bluebert. He just saw that both #Final...                  NaN   
1228  Happy Saturday here's 9 puppers on a bench. 99...                  NaN   
1254  Here's a brigade of puppers. All look very pre...                  NaN   
1274  From left to right:\nCletus, Jerome, Alejandro...                  NaN   
1351  Here is a whole flock of puppers.  60/50 I'll ...                  NaN   
1433  Happy Wednesday here's a bucket of pups. 44/40...                  NaN   
1634  Two sneaky puppers were not initially seen, mo...                  NaN   
1635  Someone help the girl is being mugged. Several...                  NaN   
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...                  NaN   
1843  Here we have an entire platoon of puppers. Tot...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
433                        NaN                        NaN   
902                        NaN                        NaN   
1120                       NaN                        NaN   
1202                       NaN                        NaN   
1228                       NaN                        NaN   
1254                       NaN                        NaN   
1274                       NaN                        NaN   
1351                       NaN                        NaN   
1433               

In [47]:
tw_archive[tw_archive['rating_numerator'] == 0 ]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
315   835152434251116546                    NaN                  NaN   
1016  746906459439529985           7.468859e+17         4.196984e+09   

                      timestamp  \
315   2017-02-24 15:40:31 +0000   
1016  2016-06-26 03:22:31 +0000   

                                                 source  \
315   <a href="http://twitter.com/download/iphone" r...   
1016  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
315   When you're so blinded by your systematic plag...                  NaN   
1016  PUPDATE: can't see any. Even if I could, I cou...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
315                        NaN                        NaN   
1016                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
315   https://twitter.com/dog_rates/status/835152434...                 0   
1016  https://twitter.com/dog_rates/status/746906459...                 0   

      rating_denominator  name doggo floofer pupper puppo  
315                   10  None  None    None   None  None  
1016                  10  None  None    None   None  None

In [46]:
len(tw_archive[tw_archive['rating_numerator'] == 0 ])

2

In [44]:
tw_archive[tw_archive['rating_numerator'] > 20]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
290   838150277551247360           8.381455e+17         2.195506e+07   
313   835246439529840640           8.352460e+17         2.625958e+07   
340   832215909146226688                    NaN                  NaN   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
695   786709082849828864                    NaN                  NaN   
763   778027034220126208                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
979   749981277374128128                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1712  680494726643068929                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2074  670842764863651840                    NaN                  NaN   

                      timestamp  \
188   2017-04-22 19:15:32 +0000   
189   2017-04-22 19:05:32 +0000   
290   2017-03-04 22:12:52 +0000   
313   2017-02-24 21:54:03 +0000   
340   2017-02-16 13:11:49 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
695   2016-10-13 23:23:56 +0000   
763   2016-09-20 00:24:34 +0000   
902   2016-07-28 01:00:57 +0000   
979   2016-07-04 15:00:45 +0000   
1120  2016-05-13 16:15:54 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1712  2015-12-25 21:06:00 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2074  2015-11-29 05:52:33 +0000   

                                                 source  \
188   <a href="http://twitter.com/download/iphone" r...   
189   <a href="http://twitter.com/download/iphone" r...   
290   <a href="http://twitter.com/download/iphone" r...   
313   <a href="http://twitter.com/download/iphone" r...   
340   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
979   <a href="https://about.twitter.com/products/tw...   
1120  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1712  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/downloa

In [45]:
len(tw_archive[tw_archive['rating_numerator'] > 20 ])

24

In [11]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [12]:
image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [16]:
tweet_status.describe()

contributors  coordinates  favorite_count  geo            id  \
count           0.0          0.0      772.000000  0.0  7.720000e+02   
mean            NaN          NaN    15541.392487  NaN  8.279373e+17   
std             NaN          NaN    15969.492791  NaN  3.317562e+16   
min             NaN          NaN        0.000000  NaN  6.747745e+17   
25%             NaN          NaN     6451.500000  NaN  7.993058e+17   
50%             NaN          NaN    12083.500000  NaN  8.253413e+17   
75%             NaN          NaN    21525.250000  NaN  8.541816e+17   
max             NaN          NaN   141014.000000  NaN  8.924206e+17   

             id_str  in_reply_to_status_id  in_reply_to_status_id_str  \
count  7.720000e+02           3.100000e+01               3.100000e+01   
mean   8.279373e+17           8.275047e+17               8.275047e+17   
std    3.317562e+16           4.594493e+16               4.594493e+16   
min    6.747745e+17           6.671522e+17               6.671522e+17   
25%    7.993058e+17           8.123772e+17               8.123772e+17   
50%    8.253413e+17           8.380855e+17               8.380855e+17   
75%    8.541816e+17           8.558601e+17               8.558601e+17   
max    8.924206e+17           8.862664e+17               8.862664e+17   

       in_reply_to_user_id  in_reply_to_user_id_str  place  \
count         3.100000e+01             3.100000e+01    0.0   
mean          5.067913e+16             5.067913e+16    NaN   
std           1.966841e+17             1.966841e+17    NaN   
min           1.185634e+07             1.185634e+07    NaN   
25%           2.987419e+07             2.987419e+07    NaN   
50%           2.918590e+08             2.918590e+08    NaN   
75%           4.196984e+09             4.196984e+09    NaN   
max           8.405479e+17             8.405479e+17    NaN   

       possibly_sensitive  possibly_sensitive_appealable  quoted_status_id  \
count               674.0                          674.0      1.900000e+01   
mean                  0.0                            0.0      8.439625e+17   
std                   0.0                            0.0      3.114828e+16   
min                   0.0                            0.0      7.888183e+17   
25%                   0.0                            0.0      8.209341e+17   
50%                   0.0                            0.0      8.460885e+17   
75%                   0.0                            0.0      8.727446e+17   
max                   0.0                            0.0      8.860534e+17   

       quoted_status_id_str  retweet_count  
count          1.900000e+01     772.000000  
mean           8.439625e+17    5400.071244  
std            3.114828e+16    5939.504067  
min            7.888183e+17       0.000000  
25%            8.209341e+17    2309.500000  
50%            8.460885e+17    3693.000000  
75%            8.727446e+17    6198.000000  
max            8.860534e+17   61288.000000

In [18]:
assert [tw_archive.duplicated() == False]

In [19]:
assert [image_predictions.duplicated() == False]

In [21]:
assert [tweet_status.id.duplicated() == False]

no duplicates found

In [22]:
image_predictions.p1.value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
cocker_spaniel                30
malamute                      30
French_bulldog                26
miniature_pinscher            23
Chesapeake_Bay_retriever      23
seat_belt                     22
German_shepherd               20
Staffordshire_bullterrier     20
Siberian_husky                20
Cardigan                      19
web_site                      19
teddy                         18
Maltese_dog                   18
Shetland_sheepdog             18
beagle                        18
Eskimo_dog                    18
Rottweiler                    17
Lakeland_terrier              17
Shih-Tzu                      17
Italian_greyhound             16
kuvasz                        16
          

In [11]:
rare_things= image_predictions.groupby('p1').filter(lambda x: len(x) < 3)

In [14]:
rare_things.sample(5)

tweet_id                                          jpg_url  \
640   681302363064414209  https://pbs.twimg.com/media/CXR44l9WcAAcG_N.jpg   
424   674063288070742018  https://pbs.twimg.com/media/CVrA-rIWEAANxwQ.jpg   
1242  747204161125646336  https://pbs.twimg.com/media/Cl6aOBhWEAALuti.jpg   
736   687102708889812993  https://pbs.twimg.com/media/CYkURJjW8AEamoI.jpg   
1361  761227390836215808  https://pbs.twimg.com/media/CpBsRleW8AEfO8G.jpg   

      img_num              p1   p1_conf  p1_dog              p2   p2_conf  \
640         1  frilled_lizard  0.326259   False     tailed_frog  0.104539   
424         1         ostrich  0.661176   False        bearskin  0.214879   
1242        2            coil  0.533699   False          dugong  0.087959   
736         1    fiddler_crab  0.992069   False           quail  0.002491   
1361        1          cougar  0.306512   False  French_bulldog  0.280802   

      p2_dog           p3   p3_conf  p3_dog  
640    False      axolotl  0.072479   False  
424    False         swab  0.064456   False  
1242   False  rain_barrel  0.039221   False  
736    False    rock_crab  0.001513   False  
1361    True        boxer  0.054523    True

In [15]:
len(rare_things)

271

In [16]:
len(image_predictions)

2075

In [148]:
tweet_status.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 772 entries, 0 to 771
Data columns (total 32 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       772 non-null datetime64[ns]
display_text_range               772 non-null object
entities                         772 non-null object
extended_entities                627 non-null object
favorite_count                   772 non-null int64
favorited                        772 non-null bool
full_text                        772 non-null object
geo                              0 non-null float64
id                               772 non-null int64
id_str                           772 non-null int64
in_reply_to_screen_name          31 non-null object
in_reply_to_status_id            31 non-null float64
in_reply_to_status_id_str        31 non-null float64
in_reply_to_user_id              31 non-null float64
in_reply_to_user_id_str          31 non-null flo

In [154]:
tweet_status['id'].sample(5)

714    782722598790725632
483    813202720496779264
10     890006608113172480
407    822489057087389700
280    838476387338051585
Name: id, dtype: int64

In [156]:
tw_archive[tw_archive.tweet_id == 782722598790725632]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
725  782722598790725632                    NaN                  NaN   

                     timestamp  \
725  2016-10-02 23:23:04 +0000   

                                                source  \
725  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
725  This is Penny. She fought a bee and the bee wo...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
725                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
725  https://twitter.com/dog_rates/status/782722598...                10   

     rating_denominator   name doggo floofer pupper puppo  
725                  10  Penny  None    None   None  None

In [158]:
tweet_status[tweet_status.id == 782722598790725632]

contributors  coordinates          created_at display_text_range  \
714           NaN          NaN 2016-10-02 23:23:04            [0, 91]   

                                              entities  \
714  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                     extended_entities  favorite_count  \
714  {'media': [{'id': 782722587017285632, 'id_str'...           18522   

     favorited                                          full_text  geo  \
714      False  This is Penny. She fought a bee and the bee wo...  NaN   

                           ...                          quoted_status  \
714                        ...                                    NaN   

     quoted_status_id quoted_status_id_str  quoted_status_permalink  \
714               NaN                  NaN                      NaN   

     retweet_count  retweeted  retweeted_status  \
714           5933      False               NaN   

                                                source truncated  \
714  <a href="http://twitter.com/download/iphone" r...     False   

                                                  user  
714  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[1 rows x 32 columns]

In [157]:
tw_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [160]:
# compare tweet_status.created_at and tw_archive.timestamp

df1 = tweet_status.filter(['id','created_at'])
df1 = df1.rename(columns={'id': 'tweet_id'})
df2 = tw_archive.filter(['tweet_id', 'timestamp'])

dfmerged = pd.merge(df1, df2, how='inner', on=['tweet_id'])

In [161]:
dfmerged.sample(5)

tweet_id          created_at                  timestamp
247  844580511645339650 2017-03-22 16:04:20  2017-03-22 16:04:20 +0000
255  842535590457499648 2017-03-17 00:38:32  2017-03-17 00:38:32 +0000
485  813172488309972993 2016-12-26 00:00:03  2016-12-26 00:00:03 +0000
314  834089966724603904 2017-02-21 17:18:39  2017-02-21 17:18:39 +0000
192  854365224396361728 2017-04-18 16:05:17  2017-04-18 16:05:17 +0000

### Assessment report

<strong>Tidiness Issues (structural issues)</strong>

<strong>Tw_archive</strong>

- Not all columns are needed for further analysis

- Replace 'None' with np.nan to indicate the missing values

-	The columns (doggo, floofer, pupper and puppo) do not need to be separated. Each dog ‎will be classified as one of these classifications. It is better to create one column for dog ‎classification that contains the values (doggo, floofer, pupper and puppo).



<b>image_predictions</b>

-   Change the names of columns type (p1, p2 and p3) to a more descriptive name. 
-	The P1 confidence interval might not be always convincing. I will set the acceptance ‎threshold at P ≥ 0.5. This might mean neglecting a lot of the data (the median is .585). ‎However, the trade off is higher confidence in the results.‎


<b>tweet_status</b>

- Rename the column id to be tweet_id to facilitate merging

- The column tweet_status.created_at is the same as the column tw_archive.timestamp




<strong>Quality issues</strong>

<b>Tw_archive</b>

-	Some columns like (in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id) are missing ‎too many values and might not be needed in this analysis.‎
-	For the column “rating_numirator” there are zero values. delete these two columns.
-	For the column “rating_numirator” there are extreme values. 
-	Delete the cells with numerator more than 20. This will not affect the data. 
-	For the column “rating_denumirator” there is a zero value. delete it. 
-	For the column “rating_denumirator” there are extreme values. 
-	Typo mistakes in the nomirator and the denominator:‎
-	Line  1202‎	numerator must be 10 not 50, and denominator must be 10 (typo 5 ‎instead of 1)‎
-	Line  1254‎	numerator must be 10 not 80, and denominator must be 10 (typo 8 ‎instead of 1)‎
-	Line  1433‎	numerator must be 11 not 44, and denominator must be 10 (typo 4 ‎instead of 1)‎
-	Line  1843‎	numerator must be 11 not 88, and denominator must be 10 (typo 8 ‎instead of 1)‎
-	It is not clear which typo mistake is it with the rest of the extreme denominator ‎values.‎



<b>image_predictions</b>

- Drop unneeded columns
- “Egyptian_cat”, is that a dog? Delete this row.
- There are some strange values in the p1 columns such as “school_bus, pillow, cartoon” ‎these cells must be checked first then deleted if not needed. 
- 271 rows (frequency less than 3) could be deleted because they are not that frequent and most probably not dogs.

<b>tweet_status</b>

-	Drop unneeded columns



<strong>Mereg all documents usig tweet_id</strong>

## Data Cleaning

## Cleaning tw_archive

In [2]:
# Reading gathered files
tw_archive = pd.read_csv('twitter-archive-enhanced.csv')
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')
tweet_status = pd.read_json('tweet_json.txt', lines = True,encoding='utf-8')

In [3]:
# create a cleaned dataframe from the archive
cleaned_archive = tw_archive.copy()

In [4]:
# test
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

The column "in_reply_to_status_id" is showing if the tweet was original or a reply to another tweet.
Drop all the rows where this column has a value that is not NaN. This shouls reveal original tweets and delete all retweets.


In [5]:
# Define: Select rows wehre the column "in_reply_to_status_id" is Null (this will exclude all the retweets)

#Code
cleaned_archive = cleaned_archive[cleaned_archive['in_reply_to_status_id'].isnull()]

In [6]:
# test
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2278 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2278 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2278 non-null object
source                        2278 non-null object
text                          2278 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2274 non-null object
rating_numerator              2278 non-null int64
rating_denominator            2278 non-null int64
name                          2278 non-null object
doggo                         2278 non-null object
floofer                       2278 non-null object
pupper                        2278 non-null object
puppo                         2278 non-null object
dtypes: float64(4), int64(3), obje

In [7]:
# Define: Dropping unneded (missing too much values) columns using drop function.

# code
cleaned_archive = cleaned_archive.drop(['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], axis=1)

In [7]:
# Test
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2356 non-null int64
timestamp             2356 non-null object
source                2356 non-null object
text                  2356 non-null object
expanded_urls         2297 non-null object
rating_numerator      2356 non-null int64
rating_denominator    2356 non-null int64
name                  2356 non-null object
doggo                 2356 non-null object
floofer               2356 non-null object
pupper                2356 non-null object
puppo                 2356 non-null object
dtypes: int64(3), object(9)
memory usage: 221.0+ KB


In [8]:
# Define: I do not need 'expanded_urls' in my analysis. use drop function
# Code
cleaned_archive = cleaned_archive.drop(['expanded_urls'], axis = 1)

In [9]:
# Test
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2278 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2278 non-null int64
timestamp             2278 non-null object
source                2278 non-null object
text                  2278 non-null object
rating_numerator      2278 non-null int64
rating_denominator    2278 non-null int64
name                  2278 non-null object
doggo                 2278 non-null object
floofer               2278 non-null object
pupper                2278 non-null object
puppo                 2278 non-null object
dtypes: int64(3), object(8)
memory usage: 213.6+ KB


In [10]:
# Define: use loc to correct the typo mistakes in specified cells

# Code

cleaned_archive.loc[cleaned_archive['rating_numerator']==50, ['rating_numerator']] = 10
cleaned_archive.loc[cleaned_archive['rating_denominator']==50, ['rating_denominator']] = 10

cleaned_archive.loc[cleaned_archive['rating_numerator']==88, ['rating_numerator']] = 11
cleaned_archive.loc[cleaned_archive['rating_denominator']==80, ['rating_denominator']] = 10

cleaned_archive.loc[cleaned_archive['rating_numerator']==80, ['rating_numerator']] = 10
cleaned_archive.loc[cleaned_archive['rating_denominator']==80, ['rating_denominator']] = 10

cleaned_archive.loc[cleaned_archive['rating_numerator']==44, ['rating_numerator']] = 11
cleaned_archive.loc[cleaned_archive['rating_denominator']==40, ['rating_denominator']] = 10



In [11]:
# test
cleaned_archive[cleaned_archive['rating_numerator'] > 20 ]

tweet_id                  timestamp  \
340   832215909146226688  2017-02-16 13:11:49 +0000   
433   820690176645140481  2017-01-15 17:52:40 +0000   
516   810984652412424192  2016-12-19 23:06:23 +0000   
695   786709082849828864  2016-10-13 23:23:56 +0000   
763   778027034220126208  2016-09-20 00:24:34 +0000   
902   758467244762497024  2016-07-28 01:00:57 +0000   
979   749981277374128128  2016-07-04 15:00:45 +0000   
1120  731156023742988288  2016-05-13 16:15:54 +0000   
1228  713900603437621249  2016-03-27 01:29:02 +0000   
1274  709198395643068416  2016-03-14 02:04:08 +0000   
1351  704054845121142784  2016-02-28 21:25:30 +0000   
1635  684222868335505415  2016-01-05 04:00:18 +0000   
1712  680494726643068929  2015-12-25 21:06:00 +0000   
1779  677716515794329600  2015-12-18 05:06:23 +0000   
2074  670842764863651840  2015-11-29 05:52:33 +0000   

                                                 source  \
340   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
979   <a href="https://about.twitter.com/products/tw...   
1120  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1712  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
2074  <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
340   RT @dog_rates: This is Logan, the Chow who liv...                75   
433   The floofs have been released I repeat the flo...                84   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                24   
695   This is Logan, the Chow who lived. He solemnly...                75   
763   This is Sophie. She's a Jubilant Bush Pupper. ...                27   
902   Why does this never happen at my front door......               165   
979   This is Atticus. He's quite simply America af....              1776   
1120  Say hello to this unbelievably well behaved sq...               204   
1228  Happy Saturday here's 9 puppers on a bench. 99...                99   
1274  From left to right:\nCletus, Jerome, Alejandro...                45   
1351  Here is a whole flock of puppers.  60/50 I'll ...                60   
1635  Someone help the girl is being mugged. Several...               121   
1712  Here we have uncovered an entire battalion of ...                26   
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...               144   
2074  After so many requests... here you go.\n\nGood...               420   

      rating_denominator     name doggo floofer  pupper puppo  
340                   10    Logan  None    None    None  None  
433                   70     None  None    None    None  None  
516                    7      Sam  None    None    None  None  
695                   10    Logan  None    None    None  None  
763                   10   Sophie  None    None  pupper  None  
902                  150     None  None    None    None  None  
979                   10  Atticus  None    None    None  None  
1120                 170     this  None    None    None  None  
1228                  90     None  None    None    None  None  
1274                  10     None  None    None    None  None  
1351                  10        a  None    None    None  None  
1635                 110     None  None    None    None  None  
1712                  10     None  None    None    None  None  
1779                 120     None  None    None    None  None

In [14]:
len(cleaned_archive[cleaned_archive['rating_numerator'] > 20 ])

15

In [15]:
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2278 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2278 non-null int64
timestamp             2278 non-null object
source                2278 non-null object
text                  2278 non-null object
rating_numerator      2278 non-null int64
rating_denominator    2278 non-null int64
name                  2278 non-null object
doggo                 2278 non-null object
floofer               2278 non-null object
pupper                2278 non-null object
puppo                 2278 non-null object
dtypes: int64(3), object(8)
memory usage: 213.6+ KB


In [16]:
# Define: exclude all extreme and zero values from the numertor and denuminator

#Code
cleaned_archive = cleaned_archive[cleaned_archive['rating_numerator'] != 0 ]
cleaned_archive = cleaned_archive[cleaned_archive['rating_denominator'] >= 10 ]
cleaned_archive = cleaned_archive[cleaned_archive['rating_numerator'] <= 20 ]
cleaned_archive = cleaned_archive[cleaned_archive['rating_denominator'] < 20 ]


In [17]:
# test
len(cleaned_archive[cleaned_archive['rating_numerator'] > 20 ])

0

In [18]:
len(cleaned_archive[cleaned_archive['rating_denominator'] < 10 ])

0

In [19]:
len(cleaned_archive[cleaned_archive['rating_denominator'] > 20 ])

0

In [20]:
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2260 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2260 non-null int64
timestamp             2260 non-null object
source                2260 non-null object
text                  2260 non-null object
rating_numerator      2260 non-null int64
rating_denominator    2260 non-null int64
name                  2260 non-null object
doggo                 2260 non-null object
floofer               2260 non-null object
pupper                2260 non-null object
puppo                 2260 non-null object
dtypes: int64(3), object(8)
memory usage: 211.9+ KB


In [21]:
# Define: replace the value 'None' with the NaN (missing value)

# Code:
cleaned_archive =  cleaned_archive.replace('None', np.nan)

In [22]:
# test
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2260 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2260 non-null int64
timestamp             2260 non-null object
source                2260 non-null object
text                  2260 non-null object
rating_numerator      2260 non-null int64
rating_denominator    2260 non-null int64
name                  1602 non-null object
doggo                 93 non-null object
floofer               10 non-null object
pupper                252 non-null object
puppo                 29 non-null object
dtypes: int64(3), object(8)
memory usage: 211.9+ KB


In [23]:
# Define: use loc to add a new column dog_class = doggo, floofer, pupper or poppo. NaN will be used if not any of the previous

# Code
cleaned_archive.loc[cleaned_archive['doggo'] == 'doggo', 'dog_class'] = 'doggo'
cleaned_archive.loc[cleaned_archive['floofer'] == 'floofer', 'dog_class'] = 'floofer'
cleaned_archive.loc[cleaned_archive['pupper'] == 'pupper', 'dog_class'] = 'pupper'
cleaned_archive.loc[cleaned_archive['puppo'] == 'puppo', 'dog_class'] = 'puppo'



In [24]:
cleaned_archive.sample(10)

tweet_id                  timestamp  \
1780  677700003327029250  2015-12-18 04:00:46 +0000   
1608  685641971164143616  2016-01-09 01:59:19 +0000   
1071  740214038584557568  2016-06-07 16:09:13 +0000   
327   833732339549220864  2017-02-20 17:37:34 +0000   
2090  670786190031921152  2015-11-29 02:07:44 +0000   
144   863907417377173506  2017-05-15 00:02:33 +0000   
404   824025158776213504  2017-01-24 22:44:42 +0000   
1144  727286334147182592  2016-05-02 23:59:09 +0000   
66    879415818425184262  2017-06-26 19:07:24 +0000   
1348  704347321748819968  2016-02-29 16:47:42 +0000   

                                                 source  \
1780  <a href="http://twitter.com/download/iphone" r...   
1608  <a href="http://twitter.com/download/iphone" r...   
1071  <a href="http://twitter.com/download/iphone" r...   
327   <a href="http://twitter.com/download/iphone" r...   
2090  <a href="http://twitter.com/download/iphone" r...   
144   <a href="http://twitter.com/download/iphone" r...   
404   <a href="http://twitter.com/download/iphone" r...   
1144  <a href="http://twitter.com/download/iphone" r...   
66    <a href="http://twitter.com/download/iphone" r...   
1348  <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
1780  This is Ralph. He's an interpretive dancer. 10...                10   
1608  This is Otis. He just passed a cop while going...                 7   
1071  This is getting incredibly frustrating. This i...                10   
327   RT @rolltidered: This is Gabby. Now requests t...                12   
2090  This is Creg. You offered him a ride to work b...                 8   
144   This is Albus. He's quite impressive at hide a...                13   
404   "I wish we were dogs" 14/10 for @BadlandsNPS h...                14   
1144  I swear to god if we get sent another Blue Mad...                 9   
66    This is Duddles. He did an attempt. 13/10 some...                13   
1348  This is Ralphson. He's very confused. Wonderin...                10   

      rating_denominator      name doggo floofer  pupper puppo dog_class  
1780                  10     Ralph   NaN     NaN     NaN   NaN       NaN  
1608                  10      Otis   NaN     NaN  pupper   NaN    pupper  
1071                  10   getting   NaN     NaN     NaN   NaN       NaN  
327                   10     Gabby   NaN     NaN     NaN   NaN       NaN  
2090                  10      Creg   NaN     NaN     NaN   NaN       NaN  
144                   10     Albus   NaN     NaN     NaN   NaN       NaN  
404                   10       NaN   NaN     NaN     NaN   NaN       NaN  
1144                  10       NaN   NaN     NaN     NaN   NaN       NaN  
66                    10   Duddles   NaN     NaN     NaN   NaN       NaN  
1348                  10  Ralphson   NaN     NaN  pupper   NaN    pupper

In [25]:
# test
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2260 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2260 non-null int64
timestamp             2260 non-null object
source                2260 non-null object
text                  2260 non-null object
rating_numerator      2260 non-null int64
rating_denominator    2260 non-null int64
name                  1602 non-null object
doggo                 93 non-null object
floofer               10 non-null object
pupper                252 non-null object
puppo                 29 non-null object
dog_class             371 non-null object
dtypes: int64(3), object(9)
memory usage: 229.5+ KB


In [26]:
# dropping unneded doggo, floofer, pupper or poppo columns
cleaned_archive = cleaned_archive.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis = 1)

In [27]:
cleaned_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2260 entries, 0 to 2355
Data columns (total 8 columns):
tweet_id              2260 non-null int64
timestamp             2260 non-null object
source                2260 non-null object
text                  2260 non-null object
rating_numerator      2260 non-null int64
rating_denominator    2260 non-null int64
name                  1602 non-null object
dog_class             371 non-null object
dtypes: int64(3), object(5)
memory usage: 158.9+ KB


### Cleaning images dataframe

In [28]:
cleaned_images = image_predictions.copy()

In [29]:
cleaned_images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [30]:
# Define: rename columns 

# Code
cleaned_images= cleaned_images.rename(columns={'p1':'Breed_probability1', 'p2':'Breed_probability2', 'p3':'Breed_probability3'})


In [31]:
# Test
cleaned_images.sample(5)

tweet_id                                          jpg_url  \
874   698178924120031232  https://pbs.twimg.com/media/CbBuBhbWwAEGH29.jpg   
1550  793120401413079041  https://pbs.twimg.com/media/CwG6zDfWcAA8jBD.jpg   
1124  727286334147182592  https://pbs.twimg.com/media/ChfXDrGUkAEAtF-.jpg   
1870  844704788403113984  https://pbs.twimg.com/media/C7j-hkSW0AIxCZC.jpg   
60    667138269671505920  https://pbs.twimg.com/media/CUImtzEVAAAZNJo.jpg   

      img_num           Breed_probability1   p1_conf  p1_dog  \
874         1     Chesapeake_Bay_retriever  0.351868    True   
1550        1           Labrador_retriever  0.724944    True   
1124        1                       bonnet  0.146440   False   
1870        1           Labrador_retriever  0.980213    True   
60          1  West_Highland_white_terrier  0.747713    True   

     Breed_probability2   p2_conf  p2_dog  Breed_probability3   p3_conf  \
874            malinois  0.207753    True  Labrador_retriever  0.154606   
1550   golden_retriever  0.169744    True              kuvasz  0.035502   
1124               sock  0.083091   False           Chihuahua  0.070554   
1870   golden_retriever  0.007012    True              beagle  0.003147   
60              Samoyed  0.243629    True          toy_poodle  0.001804   

      p3_dog  
874     True  
1550    True  
1124    True  
1870    True  
60      True

In [32]:
# Define: Exclude all rows with P1_confidence less than 0.5

# Code: 
cleaned_images = cleaned_images[cleaned_images['p1_conf'] > 0.5 ]

In [33]:
# test
assert [cleaned_images[cleaned_images['p1_conf'] > 0.5 ]]

In [34]:
# Define: use filter to select only needed columns

# Code
cleaned_images = cleaned_images.filter(['tweet_id','Breed_probability1','p1_conf'] )



In [35]:
# test 
cleaned_images.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1239 entries, 1 to 2072
Data columns (total 3 columns):
tweet_id              1239 non-null int64
Breed_probability1    1239 non-null object
p1_conf               1239 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 38.7+ KB


In [36]:
# Define: use groupby to select not rare things in the data (frequency more than 3)

# Code
cleaned_images = cleaned_images.groupby('Breed_probability1').filter(lambda x: len(x) > 3)

In [37]:
# Test
cleaned_images.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 987 entries, 2 to 2072
Data columns (total 3 columns):
tweet_id              987 non-null int64
Breed_probability1    987 non-null object
p1_conf               987 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 30.8+ KB


## Cleaning tweet_status

In [38]:
cleaned_status = tweet_status.copy()

In [39]:
cleaned_status.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 772 entries, 0 to 771
Data columns (total 32 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       772 non-null datetime64[ns]
display_text_range               772 non-null object
entities                         772 non-null object
extended_entities                627 non-null object
favorite_count                   772 non-null int64
favorited                        772 non-null bool
full_text                        772 non-null object
geo                              0 non-null float64
id                               772 non-null int64
id_str                           772 non-null int64
in_reply_to_screen_name          31 non-null object
in_reply_to_status_id            31 non-null float64
in_reply_to_status_id_str        31 non-null float64
in_reply_to_user_id              31 non-null float64
in_reply_to_user_id_str          31 non-null flo

In [40]:
# Define: rename the column id 

# Code
cleaned_status = cleaned_status.rename(columns={'id':'tweet_id'})

In [41]:
# Test
cleaned_status.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 772 entries, 0 to 771
Data columns (total 32 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       772 non-null datetime64[ns]
display_text_range               772 non-null object
entities                         772 non-null object
extended_entities                627 non-null object
favorite_count                   772 non-null int64
favorited                        772 non-null bool
full_text                        772 non-null object
geo                              0 non-null float64
tweet_id                         772 non-null int64
id_str                           772 non-null int64
in_reply_to_screen_name          31 non-null object
in_reply_to_status_id            31 non-null float64
in_reply_to_status_id_str        31 non-null float64
in_reply_to_user_id              31 non-null float64
in_reply_to_user_id_str          31 non-null flo

In [42]:
# Define: filter/select needed columns
# Code
cleaned_status = cleaned_status.filter(['tweet_id','favorite_count','retweet_count', 'source', 'user'] )

In [43]:
# test
cleaned_status.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 772 entries, 0 to 771
Data columns (total 5 columns):
tweet_id          772 non-null int64
favorite_count    772 non-null int64
retweet_count     772 non-null int64
source            772 non-null object
user              772 non-null object
dtypes: int64(3), object(2)
memory usage: 36.2+ KB


## Merging documents to form a working dataframe

In [44]:
tweeter_df = pd.merge(cleaned_archive, cleaned_images, how='outer', on=['tweet_id'])

In [45]:
tweeter_df = pd.merge(tweeter_df, cleaned_status, how = 'outer', on=['tweet_id'])

In [46]:
tweeter_df.sample(5)

tweet_id                  timestamp  \
1219  709519240576036864  2016-03-14 23:19:03 +0000   
2285  855860136149123072                        NaN   
1335  700062718104104960  2016-02-17 21:02:13 +0000   
230   846042936437604353  2017-03-26 16:55:29 +0000   
1193  711968124745228288  2016-03-21 17:30:03 +0000   

                                               source_x  \
1219  <a href="http://twitter.com/download/iphone" r...   
2285                                                NaN   
1335  <a href="http://twitter.com/download/iphone" r...   
230   <a href="http://twitter.com/download/iphone" r...   
1193  <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
1219  This is Vince. He's a Gregorian Flapjeck. Whit...               9.0   
2285                                                NaN               NaN   
1335  This is Eazy-E. He's colorful af. Must be rare...               6.0   
230   Meet Jarvis. The snow pupsets him. Officially ...              12.0   
1193  Meet Winston. He's trapped in a cup of coffee....              10.0   

      rating_denominator     name dog_class Breed_probability1  p1_conf  \
1219                10.0    Vince       NaN                NaN      NaN   
2285                 NaN      NaN       NaN                NaN      NaN   
1335                10.0     Eazy       NaN                NaN      NaN   
230                 10.0   Jarvis       NaN   golden_retriever  0.96111   
1193                10.0  Winston    pupper                NaN      NaN   

      favorite_count  retweet_count  \
1219             NaN            NaN   
2285          5141.0         1066.0   
1335             NaN            NaN   
230          16664.0         3083.0   
1193             NaN            NaN   

                                               source_y  \
1219                                                NaN   
2285  <a href="http://twitter.com/download/iphone" r...   
1335                                                NaN   
230   <a href="http://twitter.com/download/iphone" r...   
1193                                                NaN   

                                                   user  
1219                                                NaN  
2285  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1335                                                NaN  
230   {'id': 4196983835, 'id_str': '4196983835', 'na...  
1193                                                NaN

In [48]:
tweeter_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2307 entries, 0 to 2306
Data columns (total 14 columns):
tweet_id              2307 non-null int64
timestamp             2260 non-null object
source_x              2260 non-null object
text                  2260 non-null object
rating_numerator      2260 non-null float64
rating_denominator    2260 non-null float64
name                  1602 non-null object
dog_class             371 non-null object
Breed_probability1    987 non-null object
p1_conf               987 non-null float64
favorite_count        772 non-null float64
retweet_count         772 non-null float64
source_y              772 non-null object
user                  772 non-null object
dtypes: float64(5), int64(1), object(8)
memory usage: 270.4+ KB


There are some missing values in some columns. However, based on the analysis I want to conduct this dataframe is good to go.
I do not want to drop information now that I might need later.

In [47]:
# Saving df as csv
tweeter_df.to_csv('twitter_archive_master.csv')